natlib/svcd.c

```diff
@@ -0,0 +1,97 @@
+#define SVCD_SYMBOLS \
+	{ LSTRKEY( "svcd_add_service"), LFUNCVAL ( svcd_add_service )}, \
+    { LSTRKEY( "svcd_add_attribute"), LFUNCVAL ( svcd_add_attribute )},
+
+// Lua: storm.n.svcd_add_service ( svcd_id )
+// Add a new service to the service daemon
+// this must be called before SVCD.advertise()
+static int svcd_add_service( lua_State *L ) {
+
+    //TODO: Garbage collection?
+    //Get param 1 from top of stack
+    char * svc_id = (char *) luaL_checkstring(L, 1);
+
+    lua_getglobal(L, "SVCD");
+
+    lua_pushstring(L, "blsmap");
+    lua_gettable(L, 2);
+    lua_pushstring(L, svc_id);
+    lua_pushlightfunction(L, libstorm_bl_addservice);
+    lua_pushstring(L, svc_id);
+    lua_call(L, 1, 1);
+    lua_settable(L, 3);
+
+    lua_pushstring(L, "blamap");
+    lua_gettable(L, 2);
+    lua_pushstring(L, svc_id);
+    lua_newtable(L);
+    lua_settable(L, 3);
+
+    lua_pushstring(L, "manifest");
+    lua_gettable(L, 2);
+    lua_pushstring(L, svc_id);
+    lua_newtable(L);
+    lua_settable(L, 3);
+
+    lua_pushstring(L, "manifest_map");
+    lua_gettable(L, 2);
+    lua_pushstring(L, svc_id);
+    lua_newtable(L);
+    lua_settable(L, 3);
+
+    return 0;
+}
+
+// Lua: storm.n.svcd_add_attribute ( svc_id, attr_id, write_fn )
+// Add a new attribute to a service in the service daemon
+static void svcd_add_attribute( lua_State *L ) {
+
```

Comment:
You don't need to turn these into C strings, just use lua_pushvalue() to put a copy as TOS when you need it